src/node/internal/internal_http_server.ts

```diff
@@ -0,0 +1,760 @@
+    }
+
+    const { promise, resolve, reject } = Promise.withResolvers<Response>();
+
+    let streamController: StreamController | null = null;
+    const chunks: (Buffer | Uint8Array)[] = [];
+    const state: { bytesWritten: number; contentLength: number | null } = {
+      bytesWritten: 0,
+      contentLength: null,
+    };
+
+    const handleData = (event: DataWrittenEvent): void => {
+      const chunk = this.#dataFromDataWrittenEvent(event);
+      state.bytesWritten += chunk.length;
+
+      if (
+        state.contentLength !== null &&
+        state.bytesWritten > state.contentLength
+      ) {
+        const error = new Error(
+          `Content-Length mismatch: wrote ${state.bytesWritten} bytes but Content-Length header was ${state.contentLength}`
+        );
+        streamController?.error(error);
+        return;
+      }
+
+      if (streamController) {
+        streamController.enqueue(chunk);
+      } else {
+        chunks[event.index] = chunk;
+      }
+    };
+
+    this.on('_dataWritten', handleData);
+    this.once('error', reject);
+
+    this.once(
+      '_headersSent',
+      ({ statusCode, statusMessage, headers }: HeadersSentEvent) => {
+        for (const [name, value] of headers) {
+          // Optimization: Avoid unnecessary string comparison by checking length first
+          if (name.length === 14 && name.toLowerCase() === 'content-length') {
+            state.contentLength = parseInt(value, 10);
+            break;
+          }
+        }
+
+        resolve(
+          this.#toFetchResponse({
+            statusCode,
+            statusText: statusMessage,
+            sentHeaders: headers,
+            onStreamStart: (controller) => {
+              streamController = controller;
+              for (const chunk of chunks) {
+                controller.enqueue(chunk);
+              }
+              chunks.length = 0;
+            },
+            state,
+          })
+        );
+      }
+    );
+
+    this.#fetchResponse = promise;
+  }
+
+  #toFetchResponse({
+    statusCode,
+    statusText,
+    sentHeaders,
+    onStreamStart,
+    state,
+  }: {
+    statusCode: number;
+    statusText: string;
+    sentHeaders: [header: string, value: string][];
+    onStreamStart: (controller: StreamController) => void;
+    state: { bytesWritten: number; contentLength: number | null };
+  }): Response {
+    const headers = new Headers(sentHeaders);
+    let body = null;
+
+    if (this._hasBody) {
+      body = new ReadableStream<Uint8Array>({
+        start: (controller): void => {
+          onStreamStart(controller);
+          this.once('finish', () => {
+            if (
+              state.contentLength !== null &&
+              state.bytesWritten !== state.contentLength
+            ) {
+              controller.error(
+                new Error(
+                  `Content-Length mismatch: wrote ${state.bytesWritten} bytes but Content-Length header was ${state.contentLength}`
```

Comment:
Should this be a node.js-style error with a `code`?

Reply:
Removed this error since it was wrong.